File: src/main/java/org/animefoda/topawardsbackend/entities/user/UserEntity.java
package org.animefoda.topawardsbackend.entities.user;

import jakarta.persistence.*;
import lombok.Getter;
import lombok.Setter;
import org.animefoda.topawardsbackend.entities.BaseEntity;
import org.jspecify.annotations.Nullable;
import org.springframework.security.core.GrantedAuthority;
import org.springframework.security.core.authority.SimpleGrantedAuthority;
import org.springframework.security.core.userdetails.UserDetails;

import java.util.Collection;
import java.util.List;

@Entity
@Table(name = "users")
@Getter
@Setter
public class UserEntity extends BaseEntity<UserDTO> implements UserDetails {
    @Column(unique = true, nullable = false)
    private String email;

    @Column(nullable = false)
    private String password;

    @Column(name = "name", unique = true, nullable = false)
    private String name;

    @Column(name = "type",nullable = false)
    @Enumerated(EnumType.STRING)
    private UserType type;


    @Override
    public UserDTO toDTO() {
        return new UserDTO(id,email,name,type);
    }

    @Override
    public Collection<? extends GrantedAuthority> getAuthorities() {
        // AQUI ESTÁ O TRUQUE DO ADMIN
        if (this.type == UserType.ADMIN) {
            return List.of(
                    new SimpleGrantedAuthority("ROLE_ADMIN"),
                    new SimpleGrantedAuthority("ROLE_USER")
            );
        }
        return List.of(new SimpleGrantedAuthority("ROLE_USER"));
    }

    @Override
    public String getPassword() {
        return this.password;
    }

    @Override
    public String getUsername() {
        return email;
    }

    @Override
    public boolean isAccountNonExpired() { return true; }
    @Override
    public boolean isAccountNonLocked() { return true; }
    @Override
    public boolean isCredentialsNonExpired() { return true; }
    @Override
    public boolean isEnabled() { return true; }
}

File: src/main/java/org/animefoda/topawardsbackend/entities/user/UserEntity.kt
package org.animefoda.topawardsbackend.entities.user

import jakarta.persistence.*
import org.animefoda.topawardsbackend.entities.BaseEntity
import org.springframework.security.core.GrantedAuthority
import org.springframework.security.core.authority.SimpleGrantedAuthority
import org.springframework.security.core.userdetails.UserDetails

@Entity
@Table(name = "users")
open class UserEntity : BaseEntity<UserEntity, UserDTO>(), UserDetails {

    // Em Kotlin, declarar 'var' JÁ CRIA getter e setter automaticamente!
    // Não precisa de @Getter nem @Setter
    @Column(unique = true, nullable = false)
    var email: String = ""

    @Column(name = "password", nullable = false)
    var passwordHash: String = ""

    @Column(name = "name", nullable = false)
    var name: String = ""

    @Column(name = "type", nullable = false)
    @Enumerated(EnumType.STRING)
    var type: UserType = UserType.USER

    @Column(nullable = false)
    var active: Boolean = true

    // Seus métodos de lógica continuam aqui
    override fun toDTO(): UserDTO {
        // id vem da BaseEntity (se ela for Java, use getId())
        return UserDTO(id ?: 0, email, name, type)
    }

    override fun getAuthorities(): Collection<GrantedAuthority> {
        if (this.type == UserType.ADMIN) {
            return listOf(
                SimpleGrantedAuthority("ROLE_ADMIN"),
                SimpleGrantedAuthority("ROLE_USER")
            )
        }
        return listOf(SimpleGrantedAuthority("ROLE_USER"))
    }

    // Métodos obrigatórios do UserDetails
    override fun getUsername(): String = email
    override fun getPassword(): String = passwordHash
    override fun isAccountNonExpired(): Boolean = true
    override fun isAccountNonLocked(): Boolean = true
    override fun isCredentialsNonExpired(): Boolean = true
    override fun isEnabled(): Boolean = active
}